org.eclipse.jdt.ui/ui/org/eclipse/refactoring/MakeStaticRefactoring.java

```diff
@@ -0,0 +1,885 @@
+		for (TypeParameter methodTypeParam : methodTypeParameters) {
+			methodTypeParametersNames.add(methodTypeParam.getName().getIdentifier());
+		}
+		return methodTypeParametersNames;
+	}
+
+	private void addTypeParametersFromClass(RefactoringStatus status, AST ast, ASTRewrite rewrite, ITypeParameter[] classTypeParameters, ListRewrite typeParamsRewrite, Javadoc javadoc,
+			List<String> methodParameterTypes, List<String> methodTypeParametersNames) throws JavaModelException {
+		if (classTypeParameters.length != 0) {
+			for (int i= 0; i < classTypeParameters.length; i++) {
+				String[] bounds= classTypeParameters[i].getBounds();
+				TypeParameter typeParameter= ast.newTypeParameter();
+				typeParameter.setName(ast.newSimpleName(classTypeParameters[i].getElementName()));
+				for (String bound : bounds) {
+					if (bound.contains("?")) { //$NON-NLS-1$
+						//WildCardTypes are not allowed as bounds
+						status.merge(RefactoringStatus.createFatalErrorStatus(RefactoringCoreMessages.MakeStaticRefactoring_not_available_for_wildCardTypes_as_bound));
+						return;
+					} else {
+						SimpleType boundType= ast.newSimpleType(ast.newSimpleName(bound));
+						typeParameter.typeBounds().add(boundType);
+					}
+				}
+				//Check if method needs this TypeParameter (only if one or more methodParams are of this type OR method has instance usage OR an instance of parent class is used as methodParam)
+				String typeParamName= typeParameter.getName().getIdentifier();
+				String typeParamNameAsArray= typeParamName + "[]"; //$NON-NLS-1$
+				boolean paramIsNeeded= methodParameterTypes.contains(typeParamName) || methodParameterTypes.contains(typeParamNameAsArray);
+				if (fHasInstanceUsages || paramIsNeeded) {
+					//only insert if typeParam not already existing
+					if (!methodTypeParametersNames.contains(typeParameter.getName().getIdentifier())) {
+						typeParamsRewrite.insertLast(typeParameter, null);
+						addNewTypeParamsToJavaDoc(ast, rewrite, javadoc, typeParameter);
+					}
+				}
+			}
+		}
+	}
+
+	private void addNewTypeParamsToJavaDoc(AST ast, ASTRewrite rewrite, Javadoc javadoc, TypeParameter typeParameter) {
+		if (javadoc != null) {
+			//add new type params to javaDoc
+			TextElement textElement= ast.newTextElement();
+			textElement.setText("<" + typeParameter.getName().getIdentifier() + ">"); //$NON-NLS-1$ //$NON-NLS-2$
+			TagElement newParameterTag= ast.newTagElement();
+			newParameterTag.setTagName(TagElement.TAG_PARAM);
+			newParameterTag.fragments().add(textElement);
+			ListRewrite tagsRewrite= rewrite.getListRewrite(javadoc, Javadoc.TAGS_PROPERTY);
+			tagsRewrite.insertLast(newParameterTag, null);
+		}
+	}
+
+	private void deleteOverrideAnnotation(ASTRewrite rewrite) {
+		ListRewrite listRewrite= rewrite.getListRewrite(fTargetMethodDeclaration, MethodDeclaration.MODIFIERS2_PROPERTY);
+		for (Object obj : fTargetMethodDeclaration.modifiers()) {
+			if (obj instanceof org.eclipse.jdt.core.dom.MarkerAnnotation) {
```

Comment:
Forgot to mention this one in the review: In current Java, you can write:
```java
if (obj instanceof org.eclipse.jdt.core.dom.MarkerAnnotation marker) {
   // Use marker typed as MarkerAnnotation here
}
```

Reply:
Done.